Update PROTOCOL to include change data spec (#1300)
delta-io/delta: PROTOCOL.md

 - **Set of applications-specific transactions** that have been successfully committed to the table

## File Types

> A Delta table is stored within a directory and is composed of five different types of files.

Reviewer: I think we can just remove the number from here. its easy to forget to update this.

This directory format is only used to follow existing conventions and is not required by the protocol.
Actual partition values for a file must be read from the transaction log.

### Change Data Files

> Change data files are stored in a directory at the root of the table named `_change_data`, and represent the changes for the table version they are in. For data with partition values, change data files are stored within the `_change_data` directory in their respective partitions (i.e. `_change_data/part1=value1/...`). Writers can _optionally_ produce these change data files as a consequence of operations that change underlying data, like `UPDATE`, `DELETE`, and `MERGE` operations to a Delta Lake table. When available, change data readers should use the change data files instead of computing changes from the underlying data files.

Author: the reference implementation does not write change files for append-only - should we explicitly call this out here?

Reviewer 2: btw, in case you haven't seen it, here is the public design doc: https://docs.google.com/document/d/1FQSl0S0Irckim331YvcAKyxxGBhdLiJcJH6v3fEyiC0/edit#

taken from the public github issue. https://github.com/delta-io/delta/issues/1105

> should we explicitly call this out here

I think so.

In addition to the data columns, change data files contain metadata columns that identify the type of change event:

Field Name | Data Type | Description
-|-|-
_change_type|`String`| `insert`, `update_preimage` , `update_postimage`, `delete` __(1)__
_commit_version|`Long`| The Delta log or table version containing the change.
_commit_timestamp|`Timestamp`| The timestamp associated when the commit was created.

> __(1)__ `preimage` is the value before the update, postimage is the value after the update.

Reviewer 2: nit: `postimage` instead of postimage

This directory format is only used to follow existing conventions and is not required by the protocol.
Actual partition values for a file must be read from the transaction log.

### Change Data Files

> Change data files are stored in a directory at the root of the table named `_change_data`, and represent the changes for the table version they are in. For data with partition values, change data files are stored within the `_change_data` directory in their respective partitions (i.e. `_change_data/part1=value1/...`). Writers can _optionally_ produce these change data files as a consequence of operations that change underlying data, like `UPDATE`, `DELETE`, and `MERGE` operations to a Delta Lake table. Operations that only add new data should not produce separate change files. When available, change data readers should use the change data files instead of computing changes from the underlying data files.

Reviewer 3: > Operations that only add new data should not produce separate change files.

If an operation adds no change data files, it must only add new data without deleting or updating any existing data.

Author: Not clear on what the suggested change is. It appears to be a restatement of the same thing... also shouldn't the causality be thought of in the other direction?

Instead of 

> If an operation adds no change data files, it must only add new data without deleting or updating any existing data.

Isn't it:

> If an operation adds only new data, without deleting or updating any existing data, it should not produce any change data files

Author: @zsxwing let me know what you think about this last one! resolved the other comments in the latest commit otherwise.

> Change data files are stored in a directory at the root of the table named `_change_data`, and represent the changes for the table version they are in. For data with partition values, change data files are stored within the `_change_data` directory in their respective partitions (i.e. `_change_data/part1=value1/...`). Writers can _optionally_ produce these change data files as a consequence of operations that change underlying data, like `UPDATE`, `DELETE`, and `MERGE` operations to a Delta Lake table. Operations that only add new data should not produce separate change files. When available, change data readers should use the change data files instead of computing changes from the underlying data files.

Reviewer 3: ```suggestion
Change data files are stored in a directory at the root of the table named `_change_data`, and represent the changes for the table version they are in. For data with partition values, change data files are recommended to store within the `_change_data` directory in their respective partitions (i.e. `_change_data/part1=value1/...`). Writers can _optionally_ produce these change data files as a consequence of operations that change underlying data, like `UPDATE`, `DELETE`, and `MERGE` operations to a Delta Lake table. Operations that only add new data should not produce separate change files. When available, change data readers should use the change data files instead of computing changes from the underlying data files.
```

Reviewer 2: this is a `should` and I think it's valid, no? if you are only adding new data ... then you should only add `AddFile`s and should not add `AddCDCFile`s. when we go to read ... we will see there are no `AddCDCFile`s in this version ... and so we will treat all the `AddFile`s as inserts.

so, to me, this "should" suggestion makes sense.

what do you disagree with @zsxwing ?

Reviewer 3: It is okay to add a `AddCDCFile` for inserts as long as `_change_type` is set to `insert`?

Reviewer 2: It is okay. That's why this is `should` instead of `must`.

Reviewer 3: How about this?
> If an operation only add new data or remove existing data, a writer can write only data files and commit them in `AddFile`s or `RemoveFile`s without duplicating the data into change data files.

Reviewer 2: LGTM

> In addition to the data columns, change data files contain metadata columns that identify the type of change event:

Reviewer: ```suggestion
In addition to the data columns, change data files contain additional columns that identify the type of change event:
```
metadata is a loaded term as it can be confused to the log which is table metadata

Field Name | Data Type | Description
-|-|-
_change_type|`String`| `insert`, `update_preimage` , `update_postimage`, `delete` __(1)__
_commit_version|`Long`| The Delta log or table version containing the change.

> _commit_timestamp|`Timestamp`| The timestamp associated when the commit was created.

Reviewer 3: These two columns are not in data change files.

Author: Should we still mention them somewhere since they are inferred by the reader at runtime? It's not clear if this should be an actual requirement of readers...

}
```

### Add CDC File

> The `cdc` action is used to add a [file](#change-data-files) containing only the data that was changed as part of the transaction.

Reviewer 2: Add:

When CDC readers encounter a `cdc` action in a particular Delta version, they must read from that version exclusively using the `cdc` files, rather than inferring changes from `add` and `remove` actions as they do for the other type of operations

> The `cdc` action is used to add a [file](#change-data-files) containing only the data that was changed as part of the transaction. When CDC readers encounter a cdc action in a particular Delta version, they must read from that version exclusively using the cdc files, rather than inferring changes from add and remove actions as they do for the other type of operations.

Reviewer 2: nit, sorry: When CDC readers encounter a `cdc` action ...... exclusively using the `cdc` files

The schema of the `cdc` action is as follows:

Field Name | Data Type | Description
-|-|-

> path| String | A relative path to a file from the root of the table or an absolute path to a file that should be removed from the table. The path is a URI as specified by [RFC 2396 URI Generic Syntax](https://www.ietf.org/rfc/rfc2396.txt), which needs to be decoded to get the file path.

Reviewer 2: A relative path to a change data file from the root of the table or an absolute path to a change data file that should be added to the table

partitionValues| Map[String, String] | A map from partition column to value for this file. See also [Partition Value Serialization](#Partition-Value-Serialization)
size| Long | The size of this file in bytes

> dataChange | Boolean | Should always be `false` for change data because it only mirrors the effective changes of the data files

Reviewer 2: AddCDCFile doesn't have `dataChange`. it has `tags`. see https://github.com/delta-io/delta/blob/master/core/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala#L415-L420

Author: Resolved but it's interesting that the reference implementation still writes it as a consequence of `FileAction`

Reviewer 2: oops. I'm wrong here. it has `dataChange`. sorry for that @nkarpov

    "path": "_change_data/cdc-00001-c…..snappy.parquet",
    "partitionValues": {},
    "size": 1213,
    "dataChange": false
  }
}
```

#### Writer Requirements for AddCDCFile

As of [Writer Version 4](#Writer-Version-Requirements), all writers must respect the `delta.enableChangeDataFeed` configuration flag in the metadata of the table. When `delta.enableChangeDataFeed` is `true`, writers must produce the relevant `AddCDCFile`'s for any operation that changes data, as specified in [Change Data Files](#change-data-files)

#### Reader Requirements for AddCDCFile

When available, change data readers should use the `cdc` actions in a given table version instead of computing changes from the underlying data files referenced by the `add` and `remove` actions.
Specifically, to read the row-level changes made in a version, the following strategy should be used:
1. If there are `cdc` actions in this version, then read only those to get the row-level changes, and skip the remaining `add` and `remove` actions in this version.
2. Otherwise, if there are no `cdc` actions in this version, read and treat all the rows in the `add` and `remove` actions as inserted and deleted rows, respectively.

> 

Reviewer 3: ```
3. The following extra columns should also be generated

Field Name | Data Type | Description
-|-|-
_commit_version|`Long`| The table version containing the change. This can be got from the name of the Delta log file that contains actions.
_commit_timestamp|`Timestamp`| The timestamp associated when the commit was created. This can be got from the file modification time of the Delta log file that contains actions.
```

}
```

> ### Add CDC File

Reviewer: The protocol version information is not here.

Author: Added two subsections in this area to clarify writer/reader requirements as you've asked. Let me know if you'd like to move it elsewhere. Change Data Feed is unlike the other features (like Column Mapping) so there's no clear precedent on how to organize it within this doc.

> The `cdc` action is used to add a [file](#change-data-files) containing only the data that was changed as part of the transaction. When CDC readers encounter a `cdc` action in a particular Delta version, they must read from that version exclusively using the `cdc` files, rather than inferring changes from add and remove actions as they do for the other type of operations.

Reviewer: ```suggestion
The `cdc` action is used to add a [file](#change-data-files) containing only the data that was changed as part of the transaction. When CDC readers encounter a `cdc` action in a particular Delta table version, they must read the changes made in that version exclusively using the `cdc` files. If a version has no `cdc` action, then the data in `add` and `remove` actions are read as inserted and deleted rows, respectably.
```

Reviewer: just making it more explicit

The schema of the `cdc` action is as follows:

Field Name | Data Type | Description
-|-|-
path| String | A relative path to a change data file from the root of the table or an absolute path to a change data file that should be added to the table. The path is a URI as specified by [RFC 2396 URI Generic Syntax](https://www.ietf.org/rfc/rfc2396.txt), which needs to be decoded to get the file path.
partitionValues| Map[String, String] | A map from partition column to value for this file. See also [Partition Value Serialization](#Partition-Value-Serialization)
size| Long | The size of this file in bytes
tags | Map[String, String] | Map containing metadata about this file

The following is an example of `cdc` action.

```
{
  "cdc": {
    "path": "_change_data/cdc-00001-c…..snappy.parquet",
    "partitionValues": {},
    "size": 1213,

>     "dataChange": false

Author: I had it previously but Scott advised I remove it (https://github.com/delta-io/delta/pull/1300#discussion_r933646871). Just double checking...

>     "dataChange": false

Reviewer 3: We write `dataChange`. This is an example of the json string: https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/ActionSerializerSuite.scala#L198-L202

  }
}
```

#### Writer Requirements for AddCDCFile

As of [Writer Version 4](#Writer-Version-Requirements), all writers must respect the `delta.enableChangeDataFeed` configuration flag in the metadata of the table. When `delta.enableChangeDataFeed` is `true`, writers must produce the relevant `AddCDCFile`'s for any operation that changes data, as specified in [Change Data Files](#change-data-files)

#### Reader Requirements for AddCDCFile

> When available, change data readers should use the `AddCDCFile`s in a given table version instead of computing changes from the underlying data files referenced by the `add` and `remove` actions.

Reviewer: you need to add CDF in the table in section `## Writer Version Requirements` below.

Reviewer 3: CDF is already in the table. Right now we are just missing the details of CDF.

> When available, change data readers should use the `AddCDCFile`s in a given table version instead of computing changes from the underlying data files referenced by the `add` and `remove` actions.

Reviewer: also, this statement can be made more crisp. something like 
```
When reading the row-level changes made in a version, the following strategy should be used
- If there are AddCDCFile actions, then read them
- Otherwise read AddFile action (treat all rows as Inserts) and RemoveFile actions (treat all rows as deletes).

}
```

### Add CDC File

> The `cdc` action is used to add a [file](#change-data-files) containing only the data that was changed as part of the transaction. When CDC readers encounter a `cdc` action in a particular Delta table version, they must read the changes made in that version exclusively using the `cdc` files. If a version has no `cdc` action, then the data in `add` and `remove` actions are read as inserted and deleted rows, respectively.

Reviewer 3: ```suggestion
The `cdc` action is used to add a [file](#change-data-files) containing only the data that was changed as part of the transaction. When change data readers encounter a `cdc` action in a particular Delta table version, they must read the changes made in that version exclusively using the `cdc` files. If a version has no `cdc` action, then the data in `add` and `remove` actions are read as inserted and deleted rows, respectively.
```
